### PR TITLE
:sparkles: qs.js 6.15.3 parity

### PR DESCRIPTION
**Pull request by techouse**
This pull request focuses on improving how list limits and error handling are enforced during query string (`qs`) decoding, ensuring stricter compliance with Node `qs` 6.15.3 behavior. It also adds regression test coverage for edge cases and unbalanced bracket keys. The main changes include stricter enforcement of list limits (with exceptions thrown when exceeded in strict mode), unification of error handling logic, and expanded test cases to match Node `qs` behavior.

**List limit enforcement and error handling:**

* Introduced `Utils.throwListLimitExceeded()` to standardize throwing a `RangeError` when the list limit is exceeded, and replaced all previous ad hoc error messages and throws with this utility. [[1]](diffhunk://#diff-17db63ed2e34bd40b2bf89c04f997ba760e4fd75b04e6cf03f58349bd458477dR68-R90) [[2]](diffhunk://#diff-940ba04eaf5622e85c6ff07a147048e5dac3f994c630da9b3b6a32bd29eaff8fR49-L69) [[3]](diffhunk://#diff-940ba04eaf5622e85c6ff07a147048e5dac3f994c630da9b3b6a32bd29eaff8fL80-R94) [[4]](diffhunk://#diff-940ba04eaf5622e85c6ff07a147048e5dac3f994c630da9b3b6a32bd29eaff8fL468-R488)
* Updated `Utils.combine()` and all internal merging logic to throw immediately (when `throwOnLimitExceeded` is enabled) instead of returning an overflow map, for both flat comma lists and incremental list growth. [[1]](diffhunk://#diff-17db63ed2e34bd40b2bf89c04f997ba760e4fd75b04e6cf03f58349bd458477dL871-R905) [[2]](diffhunk://#diff-17db63ed2e34bd40b2bf89c04f997ba760e4fd75b04e6cf03f58349bd458477dR916-R921) [[3]](diffhunk://#diff-17db63ed2e34bd40b2bf89c04f997ba760e4fd75b04e6cf03f58349bd458477dR942-R944) [[4]](diffhunk://#diff-940ba04eaf5622e85c6ff07a147048e5dac3f994c630da9b3b6a32bd29eaff8fL331-R351) [[5]](diffhunk://#diff-940ba04eaf5622e85c6ff07a147048e5dac3f994c630da9b3b6a32bd29eaff8fL431-R451)
* Added pre-allocation checks for flat comma-separated values to reject oversized values before splitting and allocating memory, matching Node `qs` cumulative enforcement. [[1]](diffhunk://#diff-940ba04eaf5622e85c6ff07a147048e5dac3f994c630da9b3b6a32bd29eaff8fR49-L69) [[2]](diffhunk://#diff-940ba04eaf5622e85c6ff07a147048e5dac3f994c630da9b3b6a32bd29eaff8fR310)

**Behavioral and compatibility improvements:**

* Ensured that duplicate key combinations and mixed list merges enforce the cumulative list limit across all code paths, matching Node `qs` 6.15.3. [[1]](diffhunk://#diff-940ba04eaf5622e85c6ff07a147048e5dac3f994c630da9b3b6a32bd29eaff8fL331-R351) [[2]](diffhunk://#diff-940ba04eaf5622e85c6ff07a147048e5dac3f994c630da9b3b6a32bd29eaff8fR310) [[3]](diffhunk://#diff-17db63ed2e34bd40b2bf89c04f997ba760e4fd75b04e6cf03f58349bd458477dL185-R213) [[4]](diffhunk://#diff-17db63ed2e34bd40b2bf89c04f997ba760e4fd75b04e6cf03f58349bd458477dL206-R245) [[5]](diffhunk://#diff-17db63ed2e34bd40b2bf89c04f997ba760e4fd75b04e6cf03f58349bd458477dR297-R302) [[6]](diffhunk://#diff-17db63ed2e34bd40b2bf89c04f997ba760e4fd75b04e6cf03f58349bd458477dL319-R346) [[7]](diffhunk://#diff-17db63ed2e34bd40b2bf89c04f997ba760e4fd75b04e6cf03f58349bd458477dL329-R356) [[8]](diffhunk://#diff-17db63ed2e34bd40b2bf89c04f997ba760e4fd75b04e6cf03f58349bd458477dL416-R445)
* Updated error handling in tests to expect exceptions (not overflow maps) when strict list limits are exceeded.

**Regression test coverage:**

* Added comprehensive tests for unbalanced bracket keys, chunk-boundary surrogate pairs, and cyclic compaction to match Node `qs` 6.15.3 regression coverage.

**Documentation and housekeeping:**

* Updated documentation and comments to clarify the semantics of flat vs. nested list values, and documented new error handling logic. [[1]](diffhunk://#diff-940ba04eaf5622e85c6ff07a147048e5dac3f994c630da9b3b6a32bd29eaff8fR37-R38) [[2]](diffhunk://#diff-17db63ed2e34bd40b2bf89c04f997ba760e4fd75b04e6cf03f58349bd458477dL871-R905)
* Updated `CHANGELOG.md` to reflect the bug fixes, new strict error handling, and new regression test coverage.

### CODE REVIEW

File: lib/src/extensions/decode.dart [hunk 1]
```diff
@@ -44,29 +46,34 @@ extension _$Decode on QS {
...
+        int commaCount = 0;
+        int commaIndex = val.indexOf(',');
+        while (commaIndex >= 0) {
+          commaCount++;
+          if (commaCount >= options.listLimit) {
+            Utils.throwListLimitExceeded(options.listLimit);
+          }
+          commaIndex = val.indexOf(',', commaIndex + 1);
+        }
+      }
```
[the remark this answers is not in the document]
Comment (techouse) [reply lib/src/extensions/decode.dart:63]: This comment is not valid for qs 6.15.3 parity.

Upstream intentionally:

- Checks only the current comma value before splitting.
- Uses `currentArrayLength` only for non-comma values.
- Enforces cumulative duplicate growth later in Utils.combine.

Including `currentListLength` here would deviate from qs and change observable custom-decoder calls. The pre-allocation protection targets one oversized comma token, not cumulative groups.
